pkg/cli/config/config.go

```diff
@@ -0,0 +1,246 @@
+type ProfileMap map[string]Config
+
+// Settings represents a settings.json
+type Settings struct {
+	Profiles  ProfileMap `json:"profiles"`
+	Auth      AuthConfig `json:"auth"`
+	CLIConfig CLIConfig  `json:"aks-engine"`
+}
+
+// New creates a new settings file.
+func New() *Settings {
+	return &Settings{
+		Profiles: ProfileMap{},
+	}
+}
+
+// FromReader reads settings from a JSON stream and returns the resulting Settings object.
+func FromReader(r io.Reader) (*Settings, error) {
+	s := New()
+	decoder := json.NewDecoder(r)
+	if err := decoder.Decode(s); err != nil {
+		return s, fmt.Errorf("could not read from JSON stream: %v", err)
+	}
+	return s, nil
+}
+
+// UsingProfile returns a Config object based on the requested profile name, merging in from the default settings.
+//
+// If the profile name does not exist in the settings, the default configuration
+// from the settings file is returned.
+func (s *Settings) UsingProfile(name string) *Config {
+	c := &Config{
+		Auth:      s.Auth,
+		CLIConfig: s.CLIConfig,
+	}
+
+	if profile, ok := s.Profiles[name]; ok {
+		mergo.Merge(&profile, c)
+		c = &profile
+	}
+
+	return c
+}
+
+// Validate checks to make sure that the auth config is well-formed.
+func (a *AuthConfig) Validate() error {
+	clientID, _ := uuid.FromString(a.ClientID)
+	subID, _ := uuid.FromString(a.SubscriptionID)
+
+	if a.AuthMethod == "client_secret" {
+		if clientID.String() == "00000000-0000-0000-0000-000000000000" || a.ClientSecret == "" {
+			return errors.New(`--client-id and --client-secret must be specified when --auth-method="client_secret"`)
+		}
+		// try parse the UUID
+	} else if a.AuthMethod == "client_certificate" {
+		if clientID.String() == "00000000-0000-0000-0000-000000000000" || a.CertificatePath == "" || a.PrivateKeyPath == "" {
+			return errors.New(`--client-id and --certificate-path, and --private-key-path must be specified when --auth-method="client_certificate"`)
+		}
+	}
+
+	if subID.String() == "00000000-0000-0000-0000-000000000000" {
+		s, err := getSubFromAzDir(filepath.Join(helpers.GetHomeDir(), ".azure"))
+		if err != nil || s.String() == "00000000-0000-0000-0000-000000000000" {
+			return errors.New("--subscription-id is required (and must be a valid UUID)")
+		}
+		log.Infoln("No subscription provided. Using selected subscription from azure CLI:", subID.String())
+		subID = s
+	}
+
+	_, err := azure.EnvironmentFromName(a.AzureEnvironment)
+	if err != nil {
+		return errors.New("failed to parse --azure-env as a valid target Azure cloud environment")
+	}
+	return nil
+}
+
+// NewClient fetches an AKS Engine client with the given parameters
+func (a *AuthConfig) NewClient() (armhelpers.AKSEngineClient, error) {
+	var client *armhelpers.AzureClient
+	env, err := azure.EnvironmentFromName(a.AzureEnvironment)
+	if err != nil {
+		return nil, err
+	}
+	switch a.AuthMethod {
+	case "cli":
+		client, err = armhelpers.NewAzureClientWithCLI(env, a.SubscriptionID)
+	case "device":
```

Comment:
Should we actually not support "device" authentication in this new paradigm as it's been deprecated by the Azure auth team?

@tariq1890 thoughts?

Reply:
Sure I am fine with deprecating it. We can flag it as deprecated over the next one or two releases. Once that has elapsed, we can remove it compeletely. That should give users enough of a notice. We can do this in another PR.

Reply:
So the resolution here is to keep it for now?